Sources/SwiftlyCore/HTTPClient.swift

```diff
@@ -88,35 +211,103 @@ public struct SwiftlyHTTPClient {
             return release
         }
 
-        return try await self.mapGitHubTags(limit: limit, filterMap: filterMap) { page in
-            try await self.getReleases(page: page)
+        swiftOrgFiltered.sort(by: >)
+
+        return if let limit = limit {
+            Array(swiftOrgFiltered.prefix(limit))
+        } else {
+            swiftOrgFiltered
         }
     }
 
     /// Return an array of Swift snapshots that match the given filter, up to the provided
     /// limit (default unlimited).
-    ///
-    /// TODO: retrieve these directly from swift.org instead of through GitHub.
     public func getSnapshotToolchains(
+        platform: PlatformDefinition,
+        arch a: String? = nil,
+        branch: ToolchainVersion.Snapshot.Branch,
         limit: Int? = nil,
         filter: ((ToolchainVersion.Snapshot) -> Bool)? = nil
     ) async throws -> [ToolchainVersion.Snapshot] {
-        let filter = { (gh: GitHubTag) -> ToolchainVersion.Snapshot? in
-            guard let snapshot = try gh.parseSnapshot() else {
-                return nil
+        // Fall back to using GitHub API's for snapshots on branches older than 6.0
+        if case let .release(major, _) = branch, major < 6 {
+            let filter = { (gh: GitHubTag) -> ToolchainVersion.Snapshot? in
+                guard let snapshot = try gh.parseSnapshot() else {
+                    return nil
+                }
+
+                if let filter {
+                    guard filter(snapshot) else {
+                        return nil
+                    }
+                }
+
+                return snapshot
+            }
+
+            return try await self.mapGitHubTags(limit: limit, filterMap: filter) { page in
+                try await self.getTags(page: page)
             }
+        }
+
+        let arch = if a == nil {
+#if arch(x86_64)
+            "x86_64"
+#elseif arch(arm64)
+            "aarch64"
+#else
+            #error("Unsupported processor architecture")
+#endif
+        } else {
+            a!
+        }
+
+        let branchLabel = switch branch {
+        case .main:
+            "main"
+        case let .release(major, minor):
+            "\(major).\(minor)"
+        }
+
+        let platformName = if platform.name == "xcode" {
+            "macos"
```

Comment:
the inconsistency between these swift.org APIs and swift.org's download URLs here is a bit odd. Is there any chance the API would be updated to use "xcode" like the downloads do so we can avoid needing to adjust for it here?

This question also applies for the non-macOS platform names. It'd be nice if e.g. `ubuntu1804` were used rather than the pretty name.

Reply:
@shahmishal here is some feedback regarding the swift.org API. There are some inconsistencies between the new API and the toolchain download URL's / filenames. We're working around these for now in swiftly, but perhaps this can be made a little more ergonomic for other consumers of the API.